### PR TITLE
Adding NCHW support

### PR DESCRIPTION
**Pull request by squarenabla**
get_placeholder_flattened works if time_dim != 1. copy_with_feature_dim_axis is fixed. Now it works for all feature_dim_axis 0

### CODE REVIEW

**Comment by albertz** [comment]
Next time: Choose a better title for the pull request ("The changes in the TFUtil.py" doesn't tell me too much...)

**Comment by albertz** [comment]
You are changing multiple things here (`get_placeholder_flattened` and `copy_with_feature_dim_axis`). This should be multiple separate commits (and if possible also multiple separate pull requests).
Also, I explained that you should also add some test cases for these (for the old behavior/functionality which already worked before, and for the newly added functionality).

File: TFUtil.py [hunk 1]
```diff
@@ -906,10 +921,11 @@ def get_placeholder_with_specific_batch_dim_axis(self, batch_dim_axis):
   def get_placeholder_time_flattened(self):
     assert self.placeholder is not None
     assert self.have_time_axis()
-    # flatten_with_seq_len_mask only works for these two cases at the moment:
-    assert (self.time_dim_axis, self.batch_dim_axis) == (0, 1) or (self.time_dim_axis, self.batch_dim_axis) == (1, 0)
+    # flatten_with_seq_len_mask only works for these cases at the moment:
+    assert (self.time_dim_axis, self.batch_dim_axis) in [(0, 1), (1, 0), (0, 2), (2, 0)]
```
Comment (albertz) [inline TFUtil.py]: This is too specific. If you want to support any axes for batch/time, why are you checking that it must be 0,2 or 2,0?

**Comment by squarenabla** [comment]
How should I do that? I'm testing everything in the debug shell and calling Data methods with different parameters. Do you have any examples of tests?

File: TFUtil.py [hunk 2]
```diff
@@ -2688,24 +2706,33 @@ def reversed(x):
...
+def flatten_with_seq_len_mask(x, seq_lens, batch_dim_axis=0, time_dim_axis=1, time_major=False):
   """
-  :param tf.Tensor x: shape (batch,time,...s...) with time_major=False or otherwise shape (time,batch,...s....)
+  :param tf.Tensor x: shape (batch,...s..., time, ...s'...) or shape (time,...s...., batch, ...s'...)
   :param tf.Tensor seq_lens: shape (batch,) of int32
-  :param bool time_major: if the time-dim is the first dimension in x
-  :return: tensor of shape (time', ...s...) where time' = sum(seq_len) <= batch*time
+  :param int batch_dim_axis: index of batch_dim in x
+  :param int time_dim_axis: index of time_dim in x
+  :param: bool time_major: whether time axis is 0 (redundant, kept for compatibility)
```
Comment (albertz) [inline TFUtil.py]: The doc format syntax is wrong. Please check PyCharm/PEP, or just see the other code.

File: TFUtil.py [hunk 3]
```diff
@@ -2688,24 +2706,33 @@ def reversed(x):
   return y
 
 
-def flatten_with_seq_len_mask(x, seq_lens, time_major=False):
+def flatten_with_seq_len_mask(x, seq_lens, batch_dim_axis=0, time_dim_axis=1, time_major=False):
```
Comment (albertz) [inline TFUtil.py]: You added some conflicting params (`time_dim_axis`, `batch_dim_axis` are conflicting with `time_major`). You cannot have the default params like that. Keep the default params all as `None`, extend the documentation about the default param, and add checks in the code, that only either `time_major` or `batch_dim_axis` and `time_dim_axis` can be set.

**Comment by albertz** [comment]
Look at the tests in Returnn.

**Comment by albertz** [comment]
Ah, and about commit message: Keep the message in the first line limited to 80 chars.
If you want to add further description for the commit: Add another empty line after the first line, and then you can add the description.
(Btw, if you use PyCharm, it will tell you already that. Are you ignoring this?)
(When you split the commit into multiple, as I explained earlier, you probably will not have this problem anymore.)

File: TFUtil.py (hunk 1 above)
Comment (squarenabla) [reply TFUtil.py]: Actually, it supports other cases, but they have to be checked. The only requirement is that either batch_dim_axis or time_dim_axis is 0.

File: TFUtil.py (hunk 3 above)
Comment (squarenabla) [reply TFUtil.py]: Ok

File: TFUtil.py (hunk 1 above)
Comment (albertz) [reply TFUtil.py]: Then check only that.

**Comment by albertz** [comment]
As explained, please do separate pull requests for these other things (channel first for conv layer etc).
Also, please follow my comments (about e.g. commit message, etc).

**Comment by squarenabla** [comment]
I made this commits before you message, that's why they I big and have a bulky message( Can I keep old commits as they are? That will be time consuming to split them into different commits.

**Comment by albertz** [comment]
They have to be changed anyway (see all my comments). I don't see why it's difficult then to also change the commit message.

**Comment by squarenabla** [comment]
> They have to be changed anyway (see all my comments). I don't see why it's difficult then to also change the commit message.

Done

**Comment by albertz** [comment]
As I said, please make separate pull requests, and cleanup the commits.

**Pull request closed without merging**